zarr/storage.py

```diff
@@ -1879,6 +1884,157 @@ def __delitem__(self, key):
+    prefix : string
+        Location of the "directory" to use as the root of the storage hierarchy
+        within the container.
+    account_name : string
+        The Azure blob storage account name.
+    account_key : string
+        The Azure blob storage account acess key.
+    blob_service_kwargs : dictionary
+        Extra arguments to be passed into the azure blob client, for e.g. when
+        using the emulator, pass in blob_service_kwargs={'is_emulated': True}
+
+    Notes
+    -----
+    In order to use this store, you must install the Azure Blob Storage
+    https://github.com/Azure/azure-storage-python/tree/master/azure-storage-blob_ version >= 1.3.0.
+    """
+
+    def __init__(self, container, prefix, account_name=None, account_key=None,
+                 blob_service_kwargs=None):
+        self.container_name = container
+        self.prefix = normalize_storage_path(prefix)
+        self.account_name = account_name
+        self.account_key = account_key
+        if blob_service_kwargs is not None:
+            self.blob_service_kwargs = blob_service_kwargs
+        else:  # pragma: no cover
+            self.blob_service_kwargs = dict()
+        self.client = BlockBlobService(self.account_name, self.account_key,
+                                       **self.blob_service_kwargs)
+
+    # needed for pickling
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state['client']
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self.client = BlockBlobService(self.account_name, self.account_key,
+                                       **self.blob_service_kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @staticmethod
+    def _append_path_to_prefix(path, prefix):
+        return '/'.join([normalize_storage_path(prefix),
+                         normalize_storage_path(path)])
+
+    @staticmethod
+    def _strip_prefix_from_path(path, prefix):
+        # normalized things will not have any leading or trailing slashes
+        path_norm = normalize_storage_path(path)
+        prefix_norm = normalize_storage_path(prefix)
+        return path_norm[(len(prefix_norm)+1):]
+
+    def __getitem__(self, key):
+        blob_name = '/'.join([self.prefix, key])
+        try:
+            blob = self.client.get_blob_to_bytes(self.container_name, blob_name)
+            return blob.content
+        except AzureMissingResourceHttpError:
+            raise KeyError('Blob %s not found' % blob_name)
+
+    def __setitem__(self, key, value):
+        value = ensure_bytes(value)
+        blob_name = '/'.join([self.prefix, key])
+        buffer = io.BytesIO(value)
+        self.client.create_blob_from_stream(self.container_name, blob_name, buffer)
+
+    def __delitem__(self, key):
+        if self.client.exists(self.container_name, '/'.join([self.prefix, key])):
```

Comment:
This probably isn't too important given that deletion performance is usually not something people worry about, but I wonder if this could be done in a single API call. I.e., try to call `self.client.delete_blob()` without any prior exists check, then transform the error that is returned by ABS into a KeyError if the blob doesn't exist. I.e., ask forgiveness rather than permission.